core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -310,21 +311,40 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             case KOTLIN_COROUTINES:
             case SCALA_FUTURE:
                 final CompletableFuture<?> composedFuture;
+                final AtomicReference<CompletionStage<?>> upstreamStage = new AtomicReference<>();
                 if (useBlockingTaskExecutor) {
-                    composedFuture = f.thenComposeAsync(
-                            msg -> toCompletionStage(invoke(ctx, req, msg), ctx.blockingTaskExecutor()),
-                            ctx.blockingTaskExecutor());
+                    composedFuture = f.thenComposeAsync(msg -> {
+                        final CompletionStage<?> result =
+                                toCompletionStage(invoke(ctx, req, msg), ctx.blockingTaskExecutor());
+                        upstreamStage.set(result);
+                        return result;
+                    }, ctx.blockingTaskExecutor());
                 } else {
-                    composedFuture = f.thenCompose(
-                            msg -> toCompletionStage(invoke(ctx, req, msg), ctx.eventLoop()));
+                    composedFuture = f.thenCompose(msg -> {
+                        final CompletionStage<?> result =
+                                toCompletionStage(invoke(ctx, req, msg), ctx.eventLoop());
+                        upstreamStage.set(result);
+                        return result;
+                    });
                 }
-                return composedFuture.handle(
-                        (result, cause) -> {
-                            if (cause != null) {
-                                return handleExceptionWithContext(exceptionHandler, ctx, req, cause);
-                            }
-                            return convertResponse(ctx, req, null, result, HttpHeaders.of());
-                        });
+
+                final CompletableFuture<HttpResponse> resFuture = composedFuture.handle((result, cause) -> {
+                    if (cause != null) {
+                        return handleExceptionWithContext(exceptionHandler, ctx, req, cause);
+                    }
+                    return convertResponse(ctx, req, null, result, HttpHeaders.of());
+                });
+                // Propagate cancellation to the upstream.
+                resFuture.handle((ignored, cause) -> {
+                    if (cause != null) {
+                        final CompletionStage<?> upstream = upstreamStage.get();
+                        if (upstream != null) {
+                            upstream.toCompletableFuture().completeExceptionally(cause);
+                        }
+                    }
+                    return null;
+                });
```

Comment:
Could these two `handle()` calls be merged into a single call? The semantic will change slightly (`upstream.completeExceptionally()` will be invoked before `resFuture` is completed, but maybe that's fine?

Reply:
I've tried to find a way, but it seems the nature of `CompletableFuture` doesn't allow me to do that since its chaining only works from upstream to downstream.

Having `upstream.completeExceptionally()` within `composedFuture.handle(...)` and just returning `resFuture` didn't work since cancelling the `resFuture` doesn't trigger the callback attached to `composedFuture`.

Reply:
By the way, the `composedFuture` will be removed in the https://github.com/line/armeria/pull/3674/files#diff-a5b50e72b0031befe6de444cf1fc3d47c573cd73a65312068dd7ba99d440f8b1
How about preceding this PR after #3674 is merged. I will try to finish the PR as soon as possible. 💪

Reply:
@ikhoon I will, thanks for letting me know!

Reply:
@ks-yim #3674 has been merged. Could you merge master and continue this PR? :bow:

Reply:
Thank you for your patience. It's your turn. 😉

Reply:
@ikhoon @trustin Resolved the conflict, PATL! 🙏